script/tests/test_formatting_locks.sh

```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+set -euo pipefail
+source ./script/functions.sh
+REPO_DIRS=$(find ./formatting_locks -type d -mindepth 2)
+for REPO_DIR in $REPO_DIRS
+do
+    HEAD=$(cat "$REPO_DIR/HEAD")
+    FILES=$(cat "$REPO_DIR/files")
+    REPO=$(echo "$REPO_DIR" | sed 's/.\/formatting_locks\///')
+    (
+    cd "$(mktemp -d)"
+    git clone "https://github.com/$REPO" "$REPO"
+    cd "$REPO"
+    git checkout "$HEAD"
+    f_rubyfmt -i lib/
+    for FILE in $FILES
+    do
+        fn=$(echo "$FILE" | cut -f1 -d,)
+        expected_md5=$(echo "$FILE" | cut -f2 -d,)
+        res_md5=$(f_md5 < "$fn")
+        if [[ "$res_md5" != "$expected_md5" ]]
+        then
+            echo "$fn did not match"
+            exit 1
+        fi
+
+    done
+    )
+done
+echo 'everything checks out'
+
```

Comment:
The idea with this is that we recognize files in the wild that demonstrate some particular behavior of the formatter and add them to the test suite? I like it!

I'm curious though why this wouldn't reuse the mechanism already used for fixtures, when this seems like a very similar method of testing. This script would provide a lot less information in the case of a failure, ie, it would say there was a failure but not what the diff was, and it would be more work to go about finding out what the change was. The time taken to run these tests would also grow linearly with the number of repos added, since there would need to be a git checkout for each versus including the individual files in this repo.

Reply:
I've been meaning to ask how you'd feel about rewriting some of the test scripts in (probably) ruby and this seems like a decent chance to do so. I've written my fair share of bash scripts and when they are around the size of this one and the fixture testing script is when I get the feeling I should rewrite them in something like ruby.

Not entirely related, but I've been frustrated with the "fail first" style of running tests being used for the fixtures, because I don't often just want to know if anything failed but what all of the failures are. Running the tests with a full fledged test runner would make it easier to turn that bit on or off when appropriate.

Reply:
> I've been meaning to ask how you'd feel about rewriting some of the test scripts in (probably) ruby and this seems like a decent chance to do so. I've written my fair share of bash scripts and when they are around the size of this one and the fixture testing script is when I get the feeling I should rewrite them in something like ruby.
> 
> Not entirely related, but I've been frustrated with the "fail first" style of running tests being used for the fixtures, because I don't often just want to know if anything failed but what all of the failures are. Running the tests with a full fledged test runner would make it easier to turn that bit on or off when appropriate.

As a former Ruby test framework maintainer, I wholly endorse this. (please use RSpec)

Reply:
> I'm curious though why this wouldn't reuse the mechanism already used for fixtures, when this seems like a very similar method of testing. This script would provide a lot less information in the case of a failure, ie, it would say there was a failure but not what the diff was, and it would be more work to go about finding out what the change was. The time taken to run these tests would also grow linearly with the number of repos added, since there would need to be a git checkout for each versus including the individual files in this repo.

The idea here is that fixtures should be samples we've manually created and checked in to exhibit specific test cases in most places. Formatting locks are moreso designed to be a quick check against O(lots) of ruby, and so we use this mechanism so we don't have to check all that code in

Reply:
> As a former Ruby test framework maintainer, I wholly endorse this. (please use RSpec)

I was going to use minitest/spec just so there wasn't an external dependency / `bundle install` step, haha. I'll play around with using rspec and see what I come up with.

Reply:
> Formatting locks are moreso designed to be a quick check against O(lots) of ruby, and so we use this mechanism so we don't have to check all that code in

Ok, I might be missing something because of the small numbers of samples so far, so I'll trust your judgement. :+1: